File: exercises/exercises-mpi-cxx/scangather.cxx
/****************************************************************
 ****
 **** This program file is part of the book 
 **** `Parallel programming for Science and Engineering'
 **** by Victor Eijkhout, eijkhout@tacc.utexas.edu
 ****
 **** copyright Victor Eijkhout 2012-2021
 ****
 **** MPI Exercise for the use of Scan/Exscan
 ****
 ****************************************************************/

#include <iostream>
#include <iomanip>
#include <sstream>
#include <vector>
#include <cmath>
using namespace std;
#include <mpi.h>

#define max(a,b) ( (a)>(b) ? (a) : (b) )

int main(int argc,char **argv) {

  MPI_Comm comm;

  MPI_Init(&argc,&argv);
  comm = MPI_COMM_WORLD;
  int nprocs,procno;
  MPI_Comm_size(comm,&nprocs);
  MPI_Comm_rank(comm,&procno);

  stringstream proctext;

  // first set a unique random seed
  srand((int)(procno*(double)RAND_MAX/nprocs));

  /*
   * How many elements do I have locally?
   * Allocate space.
   */
  int
    my_number_of_elements = rand() % nprocs,
    my_first_index=0;
  /*
   * Exercise:
   * - determine the index of your first element.
   *   Which routine do you use here?
   */
/**** your code here ****/
    (&my_number_of_elements,&my_first_index,
     1,MPI_INT,MPI_SUM,comm);
  printf("Proc %3d has %3d elements, range [%4d,%4d)\n",
         procno,my_number_of_elements,my_first_index,my_first_index+my_number_of_elements);

  /*
   * Create a local array of size `my_number_of_elements'
   * Fill in this local array
   */
  int total_number_of_elements;
  vector<int> my_elements( my_number_of_elements );
  for (int i_element=0; i_element<my_number_of_elements; i_element++)
    my_elements[i_element] = my_first_index+i_element;

  MPI_Reduce( &my_number_of_elements,&total_number_of_elements,
	      1,MPI_INT,MPI_SUM,0,comm);
  if (procno==0) {
    printf("Total number of elements: %d\n",total_number_of_elements);
  }

  /*
   * Gather all local arrays into one global
   */
  // how many elements from each process?
  vector<int> size_buffer;
  if (procno==0)
    size_buffer = vector<int>( nprocs );
  MPI_Gather
    (
/**** your code here ****/
     );
  // where are they going to go in the big buffer?
  vector<int> displ_buffer;
  if (procno==0)
    displ_buffer = vector<int>( nprocs );
  MPI_Gather
    (
/**** your code here ****/
     );
  // now create the big buffer
  vector<int> gather_buffer;
  if (procno==0)
    gather_buffer = vector<int>( total_number_of_elements );
  /*
   * Use Gatherv to collect the small buffers into a big one
   */
  MPI_Gatherv
    (
/**** your code here ****/
     );

  /*
   * Print the gathered material
   */
  if (procno==0) {
    printf("Gathered:");
    for (int i_element=0; i_element<total_number_of_elements; i_element++)
      printf(" %d",gather_buffer[i_element]);
    printf("\n");
  }

  MPI_Finalize();
  return 0;
}

File: exercises/exercises-mpi-p/scangather.py
#!/usr/bin/env python
################################################################
################################################################
####
#### This program file is part of the book and course
####   Parallel Computing for Science and Engineering
#### by Victor Eijkhout, copyright 2013-2023
####
#### scangatherp.py : scanning and gathering
####
################################################################
################################################################

from mpi4py import MPI
import math
import numpy as np
import random
import sys

comm = MPI.COMM_WORLD

nprocs = comm.Get_size()
procno = comm.Get_rank()

##
## How many elements do I have locally?
#### Allocate space.
##
random.seed(procno)
my_number_of_elements = np.zeros(1,dtype=np.intc)
my_number_of_elements[0] = int( random.random() * nprocs )
my_first_index = np.zeros(1,dtype=np.intc)
my_first_index[0] = 0

## 
##  Exercise:
##  - determine the index of your first element.
##    Which routine do you use here?
##

## **** your code here **** ##
#comm.YourScan(my_number_of_elements,my_first_index,MPI.SUM)
print( f"Proc {procno} has {my_number_of_elements[0]} elements, range [ {my_first_index[0]},{my_first_index[0]+my_number_of_elements[0]} )" )

## 
##  Create a local array of size `my_number_of_elements'
##  Fill in this local array
##

total_number_of_elements = np.zeros(1,dtype=np.intc)
my_elements = np.zeros(my_number_of_elements,dtype=np.intc)
for i_element in range(my_number_of_elements[0]):
    my_elements[i_element] = my_first_index+i_element

comm.Reduce( my_number_of_elements,total_number_of_elements,MPI.SUM,0)
if procno==0:
    print( f"Total number of elements: {total_number_of_elements}" )

## 
##  Gather all local arrays into one global
##
## how many elements from each process?
size_buffer = np.zeros(nprocs,dtype=np.intc)
comm.Gather(
## **** your code here **** ##
   );
## where are they going to go in the big buffer?
displ_buffer = np.zeros( nprocs,dtype=np.intc )
comm.Gather(
## **** your code here **** ##
   );
## now create the big buffer
gather_buffer = np.zeros(total_number_of_elements,dtype=np.intc)

## 
##  Use Gatherv to collect the small buffers into a big one
##
comm.Gatherv(
## **** your code here **** ##
   )

## 
##  Print the gathered material
##
if procno==0:
  print(f"Gathered: {gather_buffer}")
  # for i_element in range(total_number_of_elements[0]):
  #   print(" {gather_buffer[i_element]}")


